app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+            FileInputStream encryptedHDKeyBytes = getEncryptedFileStream(keyAddr);
+
+            SecretKey secretKey = (SecretKey) keyStore.getKey(keyAddr, null);
+            boolean ivExists = new File(getFilePath(context, keyAddr + "iv")).exists();
+            byte[] iv = null;
+
+            if (ivExists)
+                iv = readBytesFromFile(getFilePath(context, keyAddr + "iv"));
+            if (iv == null || iv.length == 0)
+            {
+                keyFailure("Cannot setup wallet seed.");
+                return "";
+            }
+            Cipher outCipher = Cipher.getInstance(CIPHER_ALGORITHM);
+            final GCMParameterSpec spec = new GCMParameterSpec(128, iv);
+            outCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+            CipherInputStream cipherInputStream = new CipherInputStream(encryptedHDKeyBytes, outCipher);
+            byte[] mnemonicBytes = readBytesFromStream(cipherInputStream);
+            String mnemonic = new String(mnemonicBytes);
+
+            switch (operation)
+            {
+                case FETCH_MNEMONIC:
+                    callbackInterface.FetchMnemonic(mnemonic);
+                    break;
+                case UPGRADE_HD_KEY:
+                case SIGN_DATA:
+                    return mnemonic;
+                default:
+                    break;
+            }
+        }
+        catch (InvalidKeyException e)
+        {
+            if (e instanceof UserNotAuthenticatedException)
+            {
+                checkAuthentication(operation);
+            }
+            else
+            {
+                keyFailure(e.getMessage());
+            }
+        }
+        catch (UnrecoverableKeyException e)
+        {
+            keyFailure("Key created at different security level. Please re-import key");
+            e.printStackTrace();
+        }
+        catch (IOException | CertificateException | KeyStoreException | NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException e)
+        {
+            e.printStackTrace();
+            keyFailure(e.getMessage());
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+        }
+
+        return "";
+    }
+
+    private FileInputStream getEncryptedFileStream(String currentKey) throws FileNotFoundException
+    {
+        String encryptedHDKeyPath = getEncryptedFilePath(currentKey);
+        if (encryptedHDKeyPath != null) return new FileInputStream(encryptedHDKeyPath);
+        else throw new FileNotFoundException("No key file found");
+    }
+
+    private void createHDKey()
+    {
+        HDWallet newWallet = new HDWallet(DEFAULT_KEY_STRENGTH, "");
+        storeHDKey(newWallet, Operation.CREATE_NON_AUTHENTICATED_KEY); //create non-authenticated key initially
+    }
+
+    /**
+     * Called after an authentication event was required, and the user has completed the authentication event
+     */
+    private void importHDKey()
```

Comment:
A high level comment about the import flow. The current flow (1) restores the seed from non-auth-locked key `k0`; (2) creates auth-locked key `k1` and attempts to store the seed encrypted under that; in case the user is not authenticated, we (3) create non-auth-locked key `k2`, store seed encrypted under that, request authentication and then basically repeat the flow but with `k2` as the new auth-locked key.

Step 3 seems redundant: couldn't we instead first check if the user is authenticated, and only then create the auth-locked key `k1` (overwriting `k0` which is now safe to do)? If the user fails to authenticate, we simply issue a warning and don't to anything else.